docs/enhancements/ipxe-host-boot.md

```diff
@@ -0,0 +1,125 @@
+- Serve an iPXE boot script
+- Include the ignition in the provided initrd
+
+### Non-Goals
+
+- Support PXE (tftp based)
+- Integration with ironic (separate effort)
+- iPXE boot script customization (maybe in a follow-up if it’s useful)
+- Configure the minimal ISO to pull the rootfs from the image service
+
+## Proposal
+
+The image-service will serve the kernel, rootfs, and initrd with ignition 
+appended (see Implementation Details), while the assisted service will 
+provide an API to download the boot script. The boot script created and served 
+by the assisted service will contain URLs to the required artifacts hosted by 
+the image service.
+
+### User Stories
+
+#### Story 1
+
+As a user of assisted installer I want to boot my hosts using iPXE without 
+setting up my own http server to host artifacts.
+
+### Implementation Details/Notes/Constraints
+
+#### Ironic Integration
+
+To support integration with Ironic’s existing iPXE boot methods, the 
+discovery ignition must be appended to the initrd rather than specified as 
+a separate kernel parameter. The image service will also handle this initrd 
+manipulation when an initrd is requested for a particular infra-env.
+
+#### New Image Service APIs
+
+Image service APIs will be added to download the required boot artifacts. The 
+rootfs and kernel will be served directly from the template image. The initrd 
+endpoint will be specific to an infra-env in order to embed the ignition.
+
+- `GET /boot-artifacts/{artifact-name}?{params}`
+  - artifact-name can be either “rootfs” or “kernel”
+  - params must include “version” and “arch” which are used to identify a template image
+  - No authentication is required as we’re serving these directly from the uncustomized image
+  - Examples
+    - `/boot-artifacts/rootfs?version=4.9&arch=x86_64`
+    - `/boot-artifacts/kernel?version=4.10&arch=arm64`
+- `GET /images/{image-id}/pxe-initrd?{params}`
+  - image-id is the uuid of an infra-env in assisted-service
+  - params must include “version” and “arch”
+  - params may include “api_key” or “image_token” depending on assisted-service authentication configuration
+  - Example
+    - `/images/afe293b9-e3a0-47ed-8952-f26721626497/pxe-initrd?version=4.10&arch=x86_64&image_token={token}`
+
+#### New Assisted Service REST APIs
+
+Assisted service APIs will be added to download the ipxe boot script and to 
+retrieve a presigned url for downloading the boot script.
+
+- `GET /v2/infra-envs/{infra-env-id}/downloads/ipxe-script`
+  - Will return an iPXE boot script with artifact URLs pointing to the image service
+  - Initrd URL will be presigned
+- `GET  /v2/infra-envs/{infra-env-id}/downloads/ipxe-script-url`
+  - Will return a presigned URL for the iPXE script
+
+#### Assisted Service Kubernetes APIs
+
+A presigned download url for the iPXE boot script will be added to the infra-env 
+resource status. This URL will be set as a part of the infra-env controller reconcile loop.
+
```

Comment:
For the Ironic integration purposes, it would be somewhat easier to have the initramfs link directly on the `InfraEnv` resource since that's the resource we'll be interacting with (and that's how ISO is handled).

Otherwise we'll probably start parsing the iPXE script :)

Reply:
This is a good point. I don't want to force parsing the iPXE script on anyone.

Does it make sense to do both to ease the ZTP automation? At that point should we also provide urls to the rootfs and kernel for completeness? Any thoughts on this @filanov @avishayt @mhrivnak ?

Reply:
Do we know when we're using ZTP in the infra-env controller? Maybe we put all these options in the API as optional and provide the initramfs if we're in ZTP "mode" and the script url otherwise?

Reply:
We may as well provide URLs for all artifacts.  We could be explicit and have a property for each, or provide one URL where the user will need to specify the artifact as a query param.  Not sure which is better.

Reply:
I think it's fine to have a URL for each one. I think we should not expect users to compose URLs as part of this API.
```
status:
  ipxe:
    rootfs: https://....
    initramfs: https://....
    ...
```

Reply:
Or maybe some of these artifacts would be used outside an ipxe context? Then just call the section "artifacts" or something similar instead of "ipxe".

Reply:
providing parameters sounds complicated, if it's another filed in the status i don't think that it's a big deal

Reply:
yea, better than ipxe... then we can serve the rootfs to the minimal iso in the same manner, no?

Reply:
> yea, better than ipxe... then we can serve the rootfs to the minimal iso in the same manner, no?

Nothing about the rootfs (or this url) is specific to ipxe, this is just about how we present the URL to the user in the infra-env.

Reply:
OK, so maybe let's stick with boot artifacts:
```
status:
  bootArtifacts:
    rootfs: https://....
    initramfs: https://....
    ...
```

Reply:
there could be an equiv grubby script or other items that can use these artifacts.

Reply:
Added an example for the kube api addition.